### PR TITLE
Generate `ETag`s for `InMemory` and `LocalFileSystem` (#4879)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4879

# Rationale for this change
 

Implements ETag for InMemory using a simple monotonic counter.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/local.rs [hunk 1]
```diff
@@ -996,15 +995,32 @@ fn convert_metadata(metadata: Metadata, location: Path) -> Result<ObjectMeta> {
...
 }
 
+#[cfg(unix)]
+fn get_inode(metadata: &Metadata) -> u64 {
+    std::os::unix::fs::MetadataExt::ino(metadata)
+}
+
+#[cfg(not(unix))]
+fn get_inode(metadata: &Metadata) -> u64 {
+    0
```
Comment (tustvold) [inline object_store/src/local.rs:1021]: The inode isn't strictly necessary, it is still a pretty strong ETag without it, but on Unix systems we might as well make use of it.

A happy consequence of the way most operations create a new file and then link it into place, is that the inode is actually a pretty good indicator

File: object_store/src/lib.rs [hunk 2]
```diff
@@ -718,25 +718,41 @@ pub struct GetOptions {
...
-        last_modified: DateTime<Utc>,
-    ) -> Result<()> {
-        if let Some(date) = self.if_modified_since {
-            if last_modified <= date {
-                return Err(Error::NotModified {
-                    path: location.to_string(),
-                    source: format!("{} >= {}", date, last_modified).into(),
+    ///
+    /// <https://datatracker.ietf.org/doc/html/rfc7232#section-6>
+    fn test(&self, meta: &ObjectMeta) -> Result<()> {
```
Comment (crepererum) [inline object_store/src/lib.rs:723]:
```suggestion
    fn check_modification_conditions(&self, meta: &ObjectMeta) -> Result<()> {
```

"test" is rather non-descriptive and is easily confused w/ unit tests.

**Review by crepererum — APPROVED** [review]
Minor naming nitpick.

File: object_store/src/memory.rs [hunk 3]
```diff
@@ -80,7 +77,25 @@ impl From<Error> for super::Error {
 /// storage provider.
 #[derive(Debug, Default)]
 pub struct InMemory {
-    storage: StorageType,
+    storage: SharedStorage,
+}
+
+type Entry = (Bytes, DateTime<Utc>, usize);
```
Comment (alamb) [inline object_store/src/memory.rs:83]: I think we should document what this triple is (specifically that the `usize` is an Etag)

File: object_store/src/lib.rs [hunk 4]
```diff
@@ -718,25 +718,41 @@ pub struct GetOptions {
...
-                    source: format!("{} >= {}", date, last_modified).into(),
+    ///
+    /// <https://datatracker.ietf.org/doc/html/rfc7232#section-6>
+    fn check_preconditions(&self, meta: &ObjectMeta) -> Result<()> {
+        // The use of the invalid etag "*" means no ETag is equivalent to never matching
+        let etag = meta.e_tag.as_deref().unwrap_or("*");
+        let last_modified = meta.last_modified;
+
+        if let Some(m) = &self.if_match {
+            if m != "*" && m.split(',').map(str::trim).all(|x| x != etag) {
```
Comment (alamb) [inline object_store/src/lib.rs:729]: Can you please add an explicit mention of `*`  handling as well as support for the `,` handling to the docstrings (so someone can see it on docs.rs)?

**Review by alamb — COMMENTED** [review]
Thank you @tustvold -- the Etag feature is getting quite cool and complete. 

I had two major comments:

## etag test coverage
It seems like there is no test coverage of the etag generation functionality or did I miss that?  When I say "no coverage" I mean we could remove all this code and all the tests would keep passing

## etag documentation

Also, it seems like etags in general are not well documented in the object_store crate: https://docs.rs/object_store/latest/object_store/?search=etag

I suggest that we add documentation:
1. `etags` are a supported feature of the crate (perhaps a `Etag` section with brief introduction and example of how to use them, along with which object store adapters support them)
2. Improve https://docs.rs/object_store/latest/object_store/struct.ObjectMeta.html#structfield.e_tag to refer back to the example, and pointing at `check_preconditions`)?


I think the tests should be done prior to merge. The documentation is not strictly necessary

File: object_store/src/lib.rs [hunk 5]
```diff
@@ -1685,8 +1702,86 @@ mod tests {
         assert!(stream.next().await.is_none());
     }
 
-    // Tests TODO:
-    // GET nonexisting location (in_memory/file)
-    // DELETE nonexisting location
-    // PUT overwriting
+    #[test]
+    fn test_get_options() {
```
Comment (alamb) [inline object_store/src/lib.rs:1706]: This is really a test of `check_preconditions` isn't it?

```suggestion
    fn test_pre_conditions() {
```

File: object_store/src/lib.rs [hunk 6]
```diff
@@ -1347,33 +1380,32 @@ mod tests {
             Err(e) => panic!("{e}"),
         }
 
-        if let Some(tag) = meta.e_tag {
```
Comment (tustvold) [inline object_store/src/lib.rs:1350]: I opted to make these tests "mandatory" as now all stores support them

File: object_store/src/local.rs (hunk 1 above)
Comment (alamb) [reply object_store/src/local.rs:1024]: I see you have clarified the comment that inode is used to be more resistent to collisions -- sounds good to me

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- looks great to me

**Pull request merged**